napari/utils/action_manager.py

```diff
@@ -0,0 +1,328 @@
+        self._buttons: Dict[
+            str, List[Union[QPushButton, QtStateButton]]
+        ] = defaultdict(lambda: [])
+        self._qactions: Dict[str, QAction] = defaultdict(lambda: [])
+        self._shortcuts: Dict[str, str] = {}
+        self.context: Dict[str, Any] = {}
+        self._stack: List[str] = []
+        self._recording: bool = False
+
+    def register_action(
+        self, name, command, description, keymapprovider: KeymapProvider
+    ):
+        """
+        Register an action for future usage
+
+        An action is generally a callback associated with
+         - a name (unique),
+         - a description
+         - A keymap provider (easier for focus and backward compatibility).
+
+        Actions can then be later bound/unbound from button elements, and
+        shortcuts; and the action manager will take care of modifying the keymap
+        of instances to handle shortcuts; and UI elements to have tooltips with
+        descriptions and shortcuts;
+
+        Parameters
+        ----------
+
+        name: str
+            unique name/id of the command that can be used to refer to this
+            command
+        command: callable
+            take 0, or 1 parameter; is `instance` is not None, will be called
+            `instance` as first parameter.
+        description: str
+            Long string to describe what the command does, will be used in
+            tooltips.
+
+        """
+        # if name in self._actions:
+        #    import warnings
+
+        #    warnings.warn(
+        #        f'Warning, action {name} already exists, make sure '
+        #        'you are not overwriting an existing action',
+        #        stacklevel=2,
+        #    )
+        self._actions[name] = Action(command, description, keymapprovider)
+        self._update_shortcut_bindings(name)
+        self._update_gui_elements(name)
+
+    def _update_gui_elements(self, name):
+        """
+        Update the description and shortcuts of all the (known) gui elements.
+        """
+        if name not in self._actions:
+            return
+        buttons = self._buttons.get(name, [])
+        desc = self._actions[name].description
+
+        # update buttons with shortcut and description
+        if name in self._shortcuts:
+            sht = self._shortcuts[name]
+            sht_str = f' ({format_shortcut(sht)})'
+        else:
+            sht = ''
+            sht_str = ''
+
+        for button in buttons:
+            button.setToolTip(desc + sht_str)
+            action = self._actions[name]
+
+            try:
+                # not sure how to check whether things are connected already
+                button.clicked.disconnect(action.call(self.context))
+            except Exception:
+                pass
+            button.clicked.connect(action.call(self.context))
+        if name in self._qactions:
+            action = self._actions[name]
+            qaction = self._qactions[name]
```

Comment:
This is a significant "leak" of Qt into a non-Qt part of the code base... Do we want to split this class in two, an action manager and a QAction manager that "subscribes" to it?

Reply:
Yes, likely there are other problems with that and assigning shortcuts to qactions, this is mostly to show that 1) it works, and 2) you can modify qaction after the fact and menu do update.